Samples/runtime/simulator-with-overrides/FaultyMeasurementsSimulator.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+using Microsoft.Quantum.Simulation.Simulators;
+using Microsoft.Quantum.Simulation.Core;
+using Microsoft.Quantum.Intrinsic;
+
+namespace Microsoft.Quantum.Samples.SimulatorWithOverrides
+{
+    /// <summary>
+    /// A simulator which extends QuantumSimulator and redefines measurement operation 
+    /// to return the correct measurement result with certain probability 
+    /// and the opposite measurement result in the rest of the cases.
+    /// </summary>
+    public class FaultyMeasurementsSimulator : QuantumSimulator
+    {
+        /// <summary>
+        /// The overriding definition for operation M
+        /// </summary>
+        public class M : QSimM
+        {
+            /// <summary>
+            /// The probability with which the measurement result will be flipped.
+            /// </summary>
+            const double flipProbability = 0.1;
+
+            /// <summary>
+            /// Random numbers generator used to decide when to flip the result.
+            /// </summary>
+            private static readonly System.Random rnd = new System.Random();
+
+            /// <summary>
+            /// The X gate operation used to adjust the state to match flipped measurement results.
+            /// </summary>
+            private static IUnitary<Qubit> gateX;
+
+            public M(FaultyMeasurementsSimulator m) : base(m) { }
+
+            /// <summary>
+            /// The actual definition of what the new operation does.
+            /// </summary>
+            public override Func<Qubit, Result> Body {
+                get {
+                    // Get the original M operation to call it and process the results
+                    Func<Qubit, Result> originalMeasurementOperation = base.Body;
+
+                    // Get the X gate
+                    gateX = this.Factory.Get<IUnitary<Qubit>>(typeof(X));
+
+                    // The body of the operation is a lambda
+                    return (qubit =>
+                    {
+                        // Call the original M operation to get correct measurement results
+                        Result originalResult = originalMeasurementOperation(qubit);
+
+                        // Flip the measurement result with certain probability
+                        if (rnd.NextDouble() < flipProbability) {
+                            // Remember to adjust the state of the wave function
+                            gateX.Apply(qubit);
```

Comment:
If I understand correctly, this doesn't actually implement a faulty measurement, but rather a perfect measurement immediately preceded by a symmetric bit-flip channel. For a faulty measurement, you'd likely want something described by the POVM {η 𝟙 / 2 + (1 - η) |0⟩⟨0|, η 𝟙 / 2 + (1 - η) |1⟩⟨1|} where η ≔ `2 * flipProbability`. That POVM represents that with probability 1 - η, you get the correct measurement result, while with probability η, you get a random answer by flipping a coin instead.

Reply:
Hmm, I think I did what I wanted to do (have the system collapse to a different state than it wanted to, but report measurement result consistent with what it collapsed to) but described it incorrectly. (The variant you described is one of the things I'd like the competition participants to explore.)

I'll update the descriptions accordingly. Thank you for the clarification!

Reply:
Makes sense, either way is an interesting model after all! Happy to help, in any case.